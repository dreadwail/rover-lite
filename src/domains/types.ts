import { Action, AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { RoverSearchState } from '../state/types';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type GenericThunkAction<T, S extends RoverSearchState = RoverSearchState> = ThunkAction<
  T,
  S,
  {},
  AnyAction
>;
export type PromiseThunkAction<
  T,
  S extends RoverSearchState = RoverSearchState
> = GenericThunkAction<Promise<T>, S>;

export type MapDispatchToProps<
  TDispatchProps,
  TOwnProps = {},
  TState extends RoverSearchState = RoverSearchState,
  A extends Action = Action
> = <E>(dispatch: ThunkDispatch<TState, E, A>, ownProps: TOwnProps) => TDispatchProps;
