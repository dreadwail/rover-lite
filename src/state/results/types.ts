export interface Badge {
  readonly content: string;
  readonly imageUrl: string;
  readonly title: string;
}

export interface Listing {
  readonly thumbnailUrl: string;
  readonly rank: number;
  readonly name: string;
  readonly badges: Badge[];
  readonly tagline: string;
  readonly neighborhood: string;
  readonly city: string;
  readonly state: string;
  readonly zip: string;
  readonly repeatClientCount: number;
  readonly reviewCount: number;
  readonly ratingAverage: number;
  readonly featuredReview: string;
  readonly currencySymbol: string;
  readonly price: number;
  readonly priceUnit: string;
  readonly webUrl: string;
}

export interface Results {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly isStale: boolean;
  readonly listings: Listing[];
}
