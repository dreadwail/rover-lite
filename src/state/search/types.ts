import { SearchType } from '../searchTypes/types';

export type SearchInputs = Record<string, any>;

export interface Search {
  readonly searchType: SearchType;
  readonly latitude: number;
  readonly longitude: number;
  readonly inputs: SearchInputs;
}
