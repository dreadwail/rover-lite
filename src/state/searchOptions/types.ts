export interface OptionCategory {
  readonly categoryId: string;
  readonly rank: number;
  readonly displayName: string;
}

export interface BasicOption {
  readonly type: string;
  readonly required: boolean;
  readonly label: string;
  readonly category: OptionCategory;
  readonly rank: number;
  readonly helpText?: string;
}

export interface DateOption extends BasicOption {
  readonly type: 'date';
}

export interface Choice {
  readonly value: any;
  readonly displayName: string;
}

export interface ChoiceOption extends BasicOption {
  readonly type: 'choice';
  readonly choices: Choice[];
}

export interface MultipleChoiceOption extends BasicOption {
  readonly type: 'multiple choice';
  readonly choices: Choice[];
}

export interface IntegerOption extends BasicOption {
  readonly type: 'integer';
  readonly minValue: number;
  readonly maxValue: number;
}

export interface BooleanOption extends BasicOption {
  readonly type: 'boolean';
}

export type Option =
  | BasicOption
  | DateOption
  | ChoiceOption
  | MultipleChoiceOption
  | IntegerOption
  | BooleanOption;

export const isBooleanOption = (option: Option): option is BooleanOption =>
  option.type === 'boolean';
export const isIntegerOption = (option: Option): option is IntegerOption =>
  option.type === 'integer';
export const isMultipleChoiceOption = (option: Option): option is MultipleChoiceOption =>
  option.type === 'multiple choice';
export const isChoiceOption = (option: Option): option is ChoiceOption => option.type === 'choice';
export const isDateOption = (option: Option): option is DateOption => option.type === 'date';
export const isBasicOption = (option: Option): option is BasicOption =>
  [isBooleanOption, isIntegerOption, isChoiceOption, isMultipleChoiceOption, isDateOption].every(
    check => !check(option)
  );

export type Options = Record<string, Option>;
export type OptionsByCategoryId = Record<string, Options>;

export interface SearchOptions {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly options: Options;
}
