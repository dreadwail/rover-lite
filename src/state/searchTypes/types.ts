export type SearchType =
  | 'overnight-boarding'
  | 'overnight-traveling'
  | 'drop-in'
  | 'doggy-day-care'
  | 'dog-walking';

export type Endpoints = { [T in SearchType]?: string };

export interface SearchTypes {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly endpoints: Endpoints;
}
