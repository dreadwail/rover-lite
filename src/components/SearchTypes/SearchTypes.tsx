import React, { Component } from 'react';

import { SearchType } from '../../state/searchTypes/types';
import ReloadableFailure from '../ReloadableFailure';
import Spinner from '../Spinner';

const searchTypeNames: Record<SearchType, string> = {
  'dog-walking': 'Dog Walking',
  'doggy-day-care': 'Doggy Day Care',
  'drop-in': 'Drop-In Visits',
  'overnight-boarding': 'Dog Boarding',
  'overnight-traveling': 'House Sitting',
};

export interface SearchTypesProps {
  readonly searchType: SearchType;
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly searchTypes: SearchType[];
  fetchSearchTypes(): void;
  onSearchTypeSelected(searchType: SearchType): void;
}

export default class SearchTypes extends Component<SearchTypesProps> {
  public componentDidMount() {
    const { fetchSearchTypes } = this.props;
    fetchSearchTypes();
  }

  public render() {
    const { searchType, searchTypes, isFetching, hasFailed } = this.props;

    if (isFetching) {
      return <Spinner />;
    }

    if (hasFailed) {
      return <ReloadableFailure />;
    }

    return (
      <div className="form-group">
        <label htmlFor="service-type">Service Type</label>
        <select
          className="form-control"
          id="service-type"
          value={searchType}
          onChange={this.onChange}
        >
          {searchTypes.map(type => (
            <option key={type} value={type}>
              {searchTypeNames[type]}
            </option>
          ))}
        </select>
      </div>
    );
  }

  private readonly onChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const { onSearchTypeSelected } = this.props;
    const searchType = event.target.value as SearchType;
    onSearchTypeSelected(searchType);
  };
}
