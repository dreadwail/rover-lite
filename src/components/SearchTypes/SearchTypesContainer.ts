import { connect, MapStateToProps } from 'react-redux';

import { searchTypeChanged } from '../../domains/search/actions';
import { getSearchType } from '../../domains/search/selectors';
import { fetchSearchTypes } from '../../domains/searchTypes/actions';
import { getSearchTypes, hasFailed, isFetching } from '../../domains/searchTypes/selectors';
import { MapDispatchToProps } from '../../domains/types';
import { RoverSearchState } from '../../state/types';

import SearchTypes, { SearchTypesProps } from './SearchTypes';

export type OwnProps = {};
export type MappedStateProps = Pick<
  SearchTypesProps,
  'searchType' | 'searchTypes' | 'hasFailed' | 'isFetching'
>;
export type MappedDispatchProps = Pick<
  SearchTypesProps,
  'fetchSearchTypes' | 'onSearchTypeSelected'
>;

export const mapStateToProps: MapStateToProps<
  MappedStateProps,
  OwnProps,
  RoverSearchState
> = state => ({
  searchType: getSearchType(state),
  searchTypes: getSearchTypes(state),
  hasFailed: hasFailed(state),
  isFetching: isFetching(state),
});

export const mapDispatchToProps: MapDispatchToProps<MappedDispatchProps, OwnProps> = dispatch => ({
  fetchSearchTypes: async () => {
    await dispatch(fetchSearchTypes());
  },
  onSearchTypeSelected: async searchType => {
    dispatch(searchTypeChanged(searchType));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SearchTypes);
