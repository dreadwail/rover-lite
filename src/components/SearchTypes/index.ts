export * from './SearchTypes';
export { default } from './SearchTypesContainer';
