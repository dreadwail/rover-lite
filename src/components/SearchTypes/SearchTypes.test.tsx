import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

import { SearchType } from '../../state/searchTypes/types';

import SearchTypes, { SearchTypesProps } from './SearchTypes';

describe('SearchTypes', () => {
  let props: SearchTypesProps;
  let rendered: ShallowWrapper<SearchTypesProps>;

  beforeEach(() => {
    props = {
      searchType: 'overnight-boarding',
      isFetching: false,
      hasFailed: false,
      searchTypes: [],
      fetchSearchTypes: jest.fn().mockName('fetchSearchTypes'),
      onSearchTypeSelected: jest.fn().mockName('onSearchTypeSelected'),
    };
  });

  describe('when fetching', () => {
    beforeEach(() => {
      props = { ...props, isFetching: true };
      rendered = shallow(<SearchTypes {...props} />);
    });

    it('displays a loading spinner', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('when fetching failed', () => {
    beforeEach(() => {
      props = { ...props, isFetching: false, hasFailed: true };
      rendered = shallow(<SearchTypes {...props} />);
    });

    it('displays an error message with an option to reload', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('when fetching succeeded', () => {
    beforeEach(() => {
      const searchTypes: SearchType[] = ['overnight-traveling', 'dog-walking'];
      props = { ...props, isFetching: false, hasFailed: false, searchTypes };
      rendered = shallow(<SearchTypes {...props} />);
    });

    it('renders correctly', () => {
      expect(rendered).toMatchSnapshot();
    });

    describe('when selecting one of the search types', () => {
      beforeEach(() => {
        rendered.find('#service-type').simulate('change', { target: { value: 'dog-walking' } });
      });

      it('indicates the search type has changed', () => {
        expect(props.onSearchTypeSelected).toHaveBeenCalledWith('dog-walking');
      });
    });
  });
});
