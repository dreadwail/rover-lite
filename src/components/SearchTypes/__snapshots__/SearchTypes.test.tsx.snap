// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchTypes when fetching displays a loading spinner 1`] = `<Spinner />`;

exports[`SearchTypes when fetching failed displays an error message with an option to reload 1`] = `<ReloadableFailure />`;

exports[`SearchTypes when fetching succeeded renders correctly 1`] = `
<div
  className="form-group"
>
  <label
    htmlFor="service-type"
  >
    Service Type
  </label>
  <select
    className="form-control"
    id="service-type"
    onChange={[Function]}
    value="overnight-boarding"
  >
    <option
      key="overnight-traveling"
      value="overnight-traveling"
    >
      House Sitting
    </option>
    <option
      key="dog-walking"
      value="dog-walking"
    >
      Dog Walking
    </option>
  </select>
</div>
`;
